drivers/ukbus/pci/msix.c

```diff
@@ -0,0 +1,165 @@
+#include <uk/bus/msix.h>
+
+#include <uk/plat/common/cpu.h>
+#include <uk/bus/pci.h>
+#include <uk/intctlr.h>
+
+size_t pci_msix_table_size(struct pci_device *dev)
+{
+	uint32_t table_size;
+	if (dev->msix_cap_offset == 0)
+		return 0;
+
+	PCI_CONF_READ_HEADER(uint32_t, &table_size,
+			     dev->config_addr | dev->msix_cap_offset,
+			     MSIX_TBL_SIZE);
+	return table_size + 1;
+}
+
+uint8_t pci_msix_table_bar(struct pci_device *dev)
+{
+	uint32_t mtab;
+	if (dev->msix_cap_offset == 0)
+		return UINT8_MAX;
+
+	PCI_CONF_READ_HEADER(uint32_t, &mtab,
+			     dev->config_addr | dev->msix_cap_offset,
+			     MSIX_TABLE_BAR);
+	return mtab;
+}
+
+uint32_t pci_msix_table_offset(struct pci_device *dev)
+{
+	uint32_t table_offset;
+	if (dev->msix_cap_offset == 0)
+		return 0;
+
+	PCI_CONF_READ_HEADER(uint32_t, &table_offset,
+			     dev->config_addr | dev->msix_cap_offset,
+			     MSIX_TABLE_OFFSET);
+	return table_offset;
+}
+
+static char *pci_get_bar(struct pci_device *dev, uint16_t idx)
+{
+	uint32_t addr, bar_low, bar_high;
+	char *base;
+
+	addr = dev->config_addr;
+
+	PCI_CONF_READ_OFFSET(uint32_t, &bar_low, addr,
+			     PCI_BASE_ADDRESS_0 + idx * 4, 0, UINT32_MAX);
+	if (bar_low & 0x1) {
+		/* The spec also seems to forbid I/O space BARs for MSI-X */
+		uk_pr_err("I/O Space BARs are not supported\n");
+		return NULL;
+	}
+
+	switch ((bar_low >> 1) & 0x3) {
+	case 0x0:
+		/* 32-bit base address */
+		base = (char *)((uintptr_t)bar_low & ~0xf);
+		break;
+	case 0x2:
+		/* 64-bit base address */
+		PCI_CONF_READ_OFFSET(uint32_t, &bar_high, PCI_ENABLE_BIT | addr,
+				     PCI_BASE_ADDRESS_0 + (idx + 1) * 4, 0,
+				     UINT32_MAX);
+		base = (char *)((uintptr_t)(bar_low & ~0xf)
+				| ((uintptr_t)bar_high << 32));
+		break;
+	default:
+		return NULL;
+	}
+
+
+	return base;
+}
+
+static int pci_msix_set_entry(struct pci_device *dev, uint16_t entry_id,
+			      struct pci_msix_table_entry *entry)
+{
+	uint32_t msix_table_size, msix_bar, msix_table_offset;
+	char *base;
+	volatile struct pci_msix_table_entry *msix_table;
```

Comment:
Out of curiosity: why these variable have to be marked as volatile?

Reply:
I'll replace it with a more explicit barrier.